src/metabase/query_processor.clj

```diff
@@ -339,11 +345,13 @@
   etc.). Returns results in a format appropriate for consumption by FE client. Saves QueryExecution row in application
   DB."
   {:style/indent 1}
-  [query, options :- mbql.s/Info]
-  (process-query
-    (-> query
-        (update :info merge options)
-        (assoc-in [:middleware :userland-query?] true))))
+  ([query, options :- mbql.s/Info]
+   (process-query
+     (-> query
+         (update :info merge options)
+         (assoc-in [:middleware :userland-query?] true))))
+  ([query, options :- mbql.s/Info, data-fn]
+   (process-query-and-save-execution! (assoc query :data-fn data-fn) options)))
```

Comment:
Is there a specific design reason why you opted to add `:data-fn` args, instead of popping it into `options` (which gets merged into `query`)? Is it to make it clearer, from the function signature, that it's different from the base case (async)?

Reply:
That's a good justification, but really it's a holdover from the original PR that had a separate `ostream` parameter. Will change if that's your decision.

Reply:
Ah. I didn't look at the original PR.

I'm not sure which way to go, really. On the one hand a distinct signature makes it clear which shape of the function you're using (and what return value to expect), on the other hand it's smelly when a change requires a bunch of changes in a bunch of different files. Then again, if the arguments _and_ the return value are different, maybe it should just be a separate function.

@camsaul What would you recommend?